core/common/src/main/java/alluxio/grpc/GrpcConnectionPool.java

```diff
@@ -0,0 +1,378 @@
+    }
+  }
+
+  /**
+   * @return the singleton pool instance
+   */
+  public static GrpcConnectionPool INSTANCE() {
+    return sInstance;
+  }
+
+  /**
+   * Acquires and increases the ref-count for the {@link ManagedChannel}.
+   *
+   * @param channelKey the channel key
+   * @param conf the Alluxio configuration
+   * @return a {@link GrpcConnection}
+   */
+  public GrpcConnection acquireConnection(GrpcChannelKey channelKey, AlluxioConfiguration conf) {
+    // Get a connection key.
+    GrpcConnectionKey connectionKey = getConnectionKey(channelKey, conf);
+    // Acquire connection.
+    CountingReference<ManagedChannel> connectionRef =
+        mChannels.compute(connectionKey, (key, ref) -> {
+          boolean shutdownExistingConnection = false;
+          int existingRefCount = 0;
+          if (ref != null) {
+            // Connection exists, wait for health check.
+            if (waitForConnectionReady(ref.get(), conf)) {
+              LOG.debug("Acquiring an existing connection. ConnectionKey: {}. Ref-count: {}", key,
+                  ref.getRefCount());
+
+              return ref.reference();
+            } else {
+              // Health check failed.
+              shutdownExistingConnection = true;
+            }
+          }
+          // Existing connection should shut-down.
+          if (shutdownExistingConnection) {
+            // TODO(ggezer): Implement GrpcConnectionListener for receiving notification.
+            existingRefCount = ref.getRefCount();
+            LOG.debug("Shutting down an existing unhealthy connection. "
+                + "ConnectionKey: {}. Ref-count: {}", key, existingRefCount);
+            // Shutdown the channel forcefully as it's already unhealthy.
+            shutdownManagedChannel(ref.get(), conf);
+          }
+
+          // Create a new managed channel.
+          LOG.debug("Creating a new managed channel. ConnectionKey: {}. Ref-count:{}", key,
+              existingRefCount);
+          ManagedChannel managedChannel = createManagedChannel(channelKey, conf);
+          // Set map reference.
+          return new CountingReference(managedChannel, existingRefCount).reference();
+        });
+
+    // Wrap connection reference and the connection.
+    return new GrpcConnection(connectionKey, connectionRef.get(), conf);
+  }
+
+  /**
+   * Decreases the ref-count of the {@link ManagedChannel} for the given address. It shuts down the
+   * underlying channel if reference count reaches zero.
+   *
+   * @param connectionKey the connection key
+   * @param conf the Alluxio configuration
+   */
+  public void releaseConnection(GrpcConnectionKey connectionKey, AlluxioConfiguration conf) {
+    mChannels.compute(connectionKey, (key, ref) -> {
+      Preconditions.checkNotNull(ref, "Cannot release nonexistent connection");
+      LOG.debug("Releasing connection for: {}. Ref-count: {}", key, ref.getRefCount());
+      // Shutdown managed channel.
+      if (ref.dereference() == 0) {
```

Comment:
Is ref protected by the concurrency guarantees provided by the concurrenthashmap?

Reply:
Correct.